File: FrogJump5.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
#include<unordered_map>

using namespace std;

class Solution {
public:
    bool canCross(vector<int>& stones) {
    	
    	unordered_map<int, unordered_set<int> > m;
    	m[0].insert(0);
    	
    	for(int s : stones) {
    		for(int k : m[s]) {
    			if(k - 1 > 0) m[s + k - 1].insert(k - 1);
    			m[s + k].insert(k);
    			m[s + k + 1].insert(k + 1);
			}
		}
    	
    	return m[stones.back()].size();
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> stones;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		stones.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->canCross(stones) ? cout<<"true" : cout<<"false";
	
	return 0;
}

File: FrogJump5.py
class Solution:
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        def canCross(cur, k):
            if cur == stones[-1]:
                return True
            if (cur, k) in memo:
                return False

            for i in range(cur + k + 1, cur + k - 2, -1):
                if i > cur and i in stones and canCross(i, i - cur):
                    return True
            memo.add((cur, k))
            return False

        if len(stones) > 4:
            for i in range(4, len(stones)):
                if stones[i] > stones[i - 1] * 2:
                    return False
        memo = set()
        return canCross(0, 0)


if __name__ == '__main__':
    stones = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.canCross(stones))
